internal/provider/function_direxists.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/hashicorp/terraform-plugin-framework/function"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var _ function.Function = &DirectoryExistsFunction{}
+
+type DirectoryExistsFunction struct{}
+
+func NewDirectoryExistsFunction() function.Function {
+	return &DirectoryExistsFunction{}
+}
+
+func (f *DirectoryExistsFunction) Metadata(ctx context.Context, req function.MetadataRequest, resp *function.MetadataResponse) {
+	resp.Name = "direxists"
+}
+
+func (f *DirectoryExistsFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
+	resp.Definition = function.Definition{
+		Summary: "Determines whether a directory exists at a given path.",
+		Description: "Given a path string, will return true if the directory exists. " +
+			"This function works only with directories. If used with a regular file, FIFO, or other " +
+			"special mode, it will return an error.",
+
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:        "path",
+				Description: "Relative or absolute path to check for the existence of a directory",
+			},
+		},
+		Return: function.BoolReturn{},
+	}
+}
+
+func (f *DirectoryExistsFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+	var directoryPath string
+
+	resp.Diagnostics.Append(req.Arguments.Get(ctx, &directoryPath)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	fi, err := os.Stat(directoryPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			resp.Diagnostics.Append(resp.Result.Set(ctx, types.BoolValue(false))...)
+			return
+		} else {
+			resp.Diagnostics.AddArgumentError(0, "Error checking for directory", err.Error())
+			return
+		}
+	}
+
+	if fi.IsDir() {
+		resp.Diagnostics.Append(resp.Result.Set(ctx, types.BoolValue(true))...)
+		return
+	}
+
+	// TODO: Should we mimic the terraform core logic for printing a nice message of what the underlying file mode is?
+	// https://github.com/hashicorp/terraform/blob/8da1c006f9d7f381e644a5f0e2c25094ea29571d/internal/lang/funcs/filesystem.go#L234-L257
```

Comment:
Maybe I'm off-base, but I think a difference is that sockets, named pipes, etc. _appear_ to be files (e.g. they show up in `ls` output similar to regular files) unless inspected deeper (e.g. `ls -l`). I think its harder for directories to be confused in the same manner since many tools have stronger distinctions between files and directories in their output. We can always add this later, if desired, since it would only potentially enhance the error messaging rather than changing the function behavior.

Reply:
Makes sense to me, I'm cool with leaving this as-is. Removed the comment ⚡